### PR TITLE
feat(nat/gateway): add new private nat gateway resource

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Add new private nat gateway resource.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new private NAT gateway resource support.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/nat' TESTARGS='-run=TestAccPrivateGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/nat -v -run=TestAccPrivateGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccPrivateGateway_basic
=== PAUSE TestAccPrivateGateway_basic
=== CONT  TestAccPrivateGateway_basic
--- PASS: TestAccPrivateGateway_basic (64.57s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/nat       64.627s
```

### CODE REVIEW

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+			"subnet_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The network ID of the subnet to which the private NAT gateway belongs.",
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile("^[\u4e00-\u9fa5\\w-]*$"),
```
Comment (hmbig2) [inline huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go]:
```suggestion
					validation.StringMatch(regexp.MustCompile(`^[\u4e00-\u9fa5\w-]*$`),
```

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+					PrivateSpecTypeMedium,
+					PrivateSpecTypeLarge,
+					PrivateSpecTypeExtraLarge,
+				}, false),
+			},
+			"enterprise_project_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The ID of the enterprise project to which the private NAT gateway belongs.",
```
Comment (hmbig2) [inline huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go:89]: add forceNew

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The region where the private NAT gateway is located.",
+			},
+			"subnet_id": {
```
Comment (hmbig2) [inline huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go:46]: add forceNew

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	d.SetId(resp.ID)
+
+	return resourcePrivateGatewayRead(ctx, d, meta)
+}
+
+func flattenDownLinkVpcs(vpcs []gateways.DownLinkVpc) string {
+	if len(vpcs) < 1 {
+		return ""
+	}
+	return vpcs[0].SubnetId
```
Comment (hmbig2) [inline huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go]: you can use utils.Search()

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go:89]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go:46]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go]:
```
panic: regexp: Compile(`^[\u4e00-\u9fa5\w-]*$`): error parsing regexp: invalid escape sequence: `\u`
```
That is why I use double quotes

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go:149]: No function names `Search` in the `utils` package.

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go]: Okey, `\x{4E00}-\x{9FFC}` is fine.

File: huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_private_gateway.go]: done. The function `PathSearch` is useful.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/nat' TESTARGS='-run=TestAccPrivateGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/nat -v -run=TestAccPrivateGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccPrivateGateway_basic
=== PAUSE TestAccPrivateGateway_basic
=== CONT  TestAccPrivateGateway_basic
--- PASS: TestAccPrivateGateway_basic (64.82s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/nat       64.899s
```

File: docs/resources/nat_private_gateway.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+* `subnet_id` - (Required, String, ForceNew) Specifies the network ID of the subnet to which the private NAT gateway
+  belongs.  
+  Changing this will create a new resource.
+
+* `name` - (Required, String) Specifies the private NAT gateway name.  
+  The valid length is limited from `1` to `64`, only english letters, chinese characters, digits, hyphens (-) and
+  underscores (_) are allowed.
+
+* `spec` - (Required, String) Specifies the specification of the private NAT gateway.
```
Comment (Jason-Zhang9309) [inline docs/resources/nat_private_gateway.md]: it's optional in the code

File: docs/resources/nat_private_gateway.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+* `region` - (Optional, String, ForceNew) Specifies the region where the private NAT gateway is located.  
+  If omitted, the provider-level region will be used. Changing this will create a new resource.
+
+* `subnet_id` - (Required, String, ForceNew) Specifies the network ID of the subnet to which the private NAT gateway
+  belongs.  
+  Changing this will create a new resource.
+
+* `name` - (Required, String) Specifies the private NAT gateway name.  
+  The valid length is limited from `1` to `64`, only english letters, chinese characters, digits, hyphens (-) and
```
Comment (Jason-Zhang9309) [inline docs/resources/nat_private_gateway.md]:
```suggestion
  The valid length is limited from `1` to `64`, only English letters, Chinese characters, digits, hyphens (-) and
```

File: huaweicloud/config/endpoints.go [hunk 7]
```diff
@@ -751,6 +751,11 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		WithOutProjectID: true,
 		Product:          "NAT",
 	},
+	"natv3": {
```
Comment (Jason-Zhang9309) [inline huaweicloud/config/endpoints.go:755]: should add `nat` to multiCatalogKeys

File: docs/resources/nat_private_gateway.md (hunk 5 above)
Comment (Lance52259) [reply docs/resources/nat_private_gateway.md]: done.

File: docs/resources/nat_private_gateway.md (hunk 6 above)
Comment (Lance52259) [reply docs/resources/nat_private_gateway.md]: done.

File: huaweicloud/config/endpoints.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/config/endpoints.go:755]: done.

**Pull request merged**